### PR TITLE
[Neuron] Add an option to build with neuron

### PR DESCRIPTION
**Pull request by liangfu**
This PR adds an option that setup vLLM to build with Neuron toolchain (include neuronx-cc and transformers-neuronx).

This would help us build
```
vllm-0.2.7+neuron212
```
, where the neuron version comes out of the compiler version (neuronx-cc 2.12).

This is part of the effort to add support to accelerate LLM inference with Trainium/Inferentia (see #1866) .

### CODE REVIEW

File: requirements-neuron.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+ninja  # For faster builds.
+psutil
+ray >= 2.5.1
+pandas  # Required for Ray data.
+pyarrow  # Required for Ray data.
+sentencepiece  # Required for LLaMA tokenizer.
+numpy
+einops  # Required for phi-1_5
```
Comment (WoosukKwon) [inline requirements-neuron.txt:8]: This was actually deleted in a recent PR
```suggestion
```

File: requirements-neuron.txt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+ninja  # For faster builds.
+psutil
+ray >= 2.5.1
```
Comment (WoosukKwon) [inline requirements-neuron.txt:3]: Do we use ray for distributed inference on Inferentia/Trainium? Otherwise, skipping it would not raise import errors.

File: setup.py [hunk 3]
```diff
@@ -319,6 +363,6 @@ def get_requirements() -> List[str]:
     python_requires=">=3.8",
     install_requires=get_requirements(),
     ext_modules=ext_modules,
-    cmdclass={"build_ext": BuildExtension},
+    cmdclass={} if _is_neuron() else {"build_ext": BuildExtension},
```
Comment (WoosukKwon) [inline setup.py:366]: nit: Can we change to this since CUDA is more common?
```suggestion
    cmdclass={"build_ext": BuildExtension} if not _is_neuron() else {},
```

File: vllm/model_executor/layers/activation.py [hunk 4]
```diff
@@ -30,6 +29,8 @@ def _forward(self, x: torch.Tensor) -> torch.Tensor:
         return F.silu(x[..., :d]) * x[..., d:]
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
+        from vllm._C import ops
+
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/activation.py:33]: To my understanding, this change is needed because we do not build the custom ops for NeuronX, right?

I think this change is a bit hacky and error prone. Instead, can we just not load these modules at all when using the NeuronX backend? While I need more information on your future PRs, I think this should be possible as you actually just convert the model using `transformers-neuronx` while model loading.

File: setup.py [hunk 5]
```diff
@@ -87,6 +96,28 @@ def get_hipcc_rocm_version():
...
+def get_neuronxcc_version():
+    # Run the neuronx-cc --version command
+    result = subprocess.run(['neuronx-cc', '--version'],
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.STDOUT,
+                            text=True)
+
+    # Check if the command was executed successfully
+    if result.returncode != 0:
+        print("Error running 'neuronx-cc --version'")
```
Comment (WoosukKwon) [inline setup.py:108]: Why don't we just raise an error in this case?

**Review by WoosukKwon — COMMENTED** [review]
Hi @liangfu, apologies for the late review and thanks for the PR! I like this PR in that you didn't submit a big PR at once but instead split it into small parts. :)

Overall, I think moving the import statements is not a good idea. Considering the architecture you showed last time, I think we can just skip loading the modules that try to import custom ops. WDYT?

File: vllm/model_executor/layers/activation.py (hunk 4 above)
Comment (liangfu) [reply vllm/model_executor/layers/activation.py:33]: > this change is needed because we do not build the custom ops for NeuronX, right?

You're right. There isn't a `nvcc` for compiling the CUDA code. Therefore, `vllm._C.ops` would not exist at all.

The intention for moving the import statement is to narrow down the scope, so that we only import these modules only when it is needed. Otherwise, without building `vllm._C` we cannot even 
```python
import vllm
```

> can we just not load these modules at all when using the NeuronX backend?

Can you elaborate on how can we **not** load these modules at all?

File: setup.py (hunk 5 above)
Comment (liangfu) [reply setup.py:108]: I was trying to be consistent with how we handle `hipcc --version` failure.

https://github.com/vllm-project/vllm/blob/a1b9cb2a3469e4e682e741af9a0f91e16923205d/setup.py#L77

File: requirements-neuron.txt (hunk 2 above)
Comment (liangfu) [reply requirements-neuron.txt:3]: For now, we are not using ray for distributed inference on Inferentia/Trainium.

File: vllm/model_executor/layers/activation.py (hunk 4 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/activation.py:33]: I mean, we can lazily load the modules in `model_executor` that try to import the customs ops and are not needed for the NeuronX backend. I'm willing to make this change, but need more information about your next PRs.

Comment (liangfu) [reply vllm/model_executor/layers/activation.py:33]: Happy new year, and sorry for the late reply. The follow-up PRs will include the changes:
- Configure transformers-neuronx to enable continuous batching feature in vLLM model loader.
- Compile the model after loading weights.
- Execute model with transformers-neuronx.
Some examples are provided in https://github.com/vllm-project/vllm/issues/1866 .

Regarding to the code structure change, i think we can maintain a separate folder for all supported model on Neuron devices. But i'm open for accepting better ideas.

There might be some changes on the scheduler, but i'm trying to minimize (and even eliminate) the changes on the scheduler.

File: requirements-neuron.txt (hunk 2 above)
Comment (liangfu) [reply requirements-neuron.txt]: removed ray from neuronx

File: vllm/model_executor/layers/activation.py (hunk 4 above)
Comment (liangfu) [reply vllm/model_executor/layers/activation.py]: undo the changes for all files under `model_executor/layers`

File: setup.py (hunk 5 above)
Comment (liangfu) [reply setup.py]: raising a RuntimeError

File: setup.py (hunk 3 above)
Comment (WoosukKwon) [reply setup.py:368]: nit: WDYT about this change?

File: vllm/utils.py [hunk 6]
```diff
@@ -36,6 +34,8 @@ def is_hip() -> bool:
 
 def get_max_shared_memory_bytes(gpu: int = 0) -> int:
     """Returns the maximum shared memory per thread block in bytes."""
+    from vllm._C import cuda_utils
```
Comment (WoosukKwon) [inline vllm/utils.py:37]: nit: Can we add a small comment on this so that we won't accidentally revert the change?
```suggestion
    # NOTE: This import statement should be executed lazily since
    # the Neuron-X backend does not have the `cuda_utils` module.
    from vllm._C import cuda_utils
```

**Review by WoosukKwon — APPROVED** [review]
@liangfu Apologies for the late review and thanks for addressing my comments! Left some very minor comments on styles. Looking forward to the next PRs!
